In [108]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [109]:
data_car=pd.read_csv("/content/used_cars_data.csv")

In [110]:
data_car.drop(columns=(["Unnamed: 0"]),inplace=True)

## Eda

In [111]:
data_car

brand     model  price (eur)  \
0          SEAT     Ibiza         8990   
1       Hyundai       i30         9990   
2           BMW   Serie 5        13490   
3    Volkswagen      Golf        24990   
4          Opel     Corsa        10460   
..          ...       ...          ...   
787  Volkswagen      Golf        13990   
788         Kia  Sportage        24990   
789      Abarth       500        17990   
790  Volkswagen    Tiguan        14990   
791       Dacia    Dokker        14390   

                                       engine  year  mileage (kms)      fuel  \
0                       SC 1.2 TSI 90cv Style  2016          67000  Gasolina   
1                        1.6 CRDi 110cv Tecno  2014         104868    Diésel   
2                                530d Touring  2011         137566    Diésel   
3                   GTI 2.0 TSI 169kW (230CV)  2018          44495  Gasolina   
4                        1.4 Expression 90 CV  2016          69800  Gasolina   
..                                        ...   ...            ...       ...   
787                 Edition 1.6 TDI 110CV BMT  2016          84040    Diésel   
788            1.6 GDi 97kW (132CV) Basic 4x2  2018          65872  Gasolina   
789  1.4 16v T-Jet 595 118kW (160CV) Pista E6  2019          28830  Gasolina   
790  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech  2014         162895    Diésel   
791            SL Trotamundos dCi 66kW (90CV)  2018           9800    Diésel   

        gearbox    location  
0        Manual  Granollers  
1        Manual  Viladecans  
2    Automatica  Viladecans  
3        Manual  Viladecans  
4        Manual  Sabadell 1  
..          ...         ...  
787      Manual        Gavá  
788      Manual  Viladecans  
789      Manual      Mataró  
790  Automatica      Mataró  
791      Manual    Badalona  

[792 rows x 9 columns]

### Transformation data

In [112]:
def transformation_Euro_to_Rp(currency):
  # Use currency in 17,02-2023 
  # 1 euro is Rp 16.157
  currency=16157*currency
  return currency

In [113]:
data_car["Price_Rp"]=data_car["price (eur)"].apply(transformation_Euro_to_Rp)

In [114]:
data_car

brand     model  price (eur)  \
0          SEAT     Ibiza         8990   
1       Hyundai       i30         9990   
2           BMW   Serie 5        13490   
3    Volkswagen      Golf        24990   
4          Opel     Corsa        10460   
..          ...       ...          ...   
787  Volkswagen      Golf        13990   
788         Kia  Sportage        24990   
789      Abarth       500        17990   
790  Volkswagen    Tiguan        14990   
791       Dacia    Dokker        14390   

                                       engine  year  mileage (kms)      fuel  \
0                       SC 1.2 TSI 90cv Style  2016          67000  Gasolina   
1                        1.6 CRDi 110cv Tecno  2014         104868    Diésel   
2                                530d Touring  2011         137566    Diésel   
3                   GTI 2.0 TSI 169kW (230CV)  2018          44495  Gasolina   
4                        1.4 Expression 90 CV  2016          69800  Gasolina   
..                                        ...   ...            ...       ...   
787                 Edition 1.6 TDI 110CV BMT  2016          84040    Diésel   
788            1.6 GDi 97kW (132CV) Basic 4x2  2018          65872  Gasolina   
789  1.4 16v T-Jet 595 118kW (160CV) Pista E6  2019          28830  Gasolina   
790  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech  2014         162895    Diésel   
791            SL Trotamundos dCi 66kW (90CV)  2018           9800    Diésel   

        gearbox    location   Price_Rp  
0        Manual  Granollers  145251430  
1        Manual  Viladecans  161408430  
2    Automatica  Viladecans  217957930  
3        Manual  Viladecans  403763430  
4        Manual  Sabadell 1  169002220  
..          ...         ...        ...  
787      Manual        Gavá  226036430  
788      Manual  Viladecans  403763430  
789      Manual      Mataró  290664430  
790  Automatica      Mataró  242193430  
791      Manual    Badalona  232499230  

[792 rows x 10 columns]

### view car based on  gearbox

In [115]:
data_type_car_gearbox=data_car[["brand","gearbox"]]

In [116]:
data_type_car_gearbox.groupby(by=["brand"])["gearbox"].value_counts()

brand       gearbox   
Abarth      Manual         6
            Automatica     3
Alfa        Manual         3
            Automatica     1
Audi        Manual        23
                          ..
Toyota      Manual        17
Volkswagen  Manual        28
            Automatica    23
Volvo       Manual         3
            Automatica     1
Name: gearbox, Length: 62, dtype: int64

In [117]:
data_type_car_gearbox

brand     gearbox
0          SEAT      Manual
1       Hyundai      Manual
2           BMW  Automatica
3    Volkswagen      Manual
4          Opel      Manual
..          ...         ...
787  Volkswagen      Manual
788         Kia      Manual
789      Abarth      Manual
790  Volkswagen  Automatica
791       Dacia      Manual

[792 rows x 2 columns]

In [118]:
data_car["fuel"].unique()

array(['Gasolina', 'Diésel', 'Híbrido', 'GLP', 'Eléctrico'], dtype=object)

In [119]:
def crosstab_function(df, var):
    tab = pd.DataFrame(pd.crosstab(df[var], df["gearbox"], margins=True)).reset_index()
    # tab['Percentage'] = tab[1] / tab['All'] * 100
    tab.columns = [var, 'Automatica', 'Manual','All']
    return tab

In [120]:
data_brand_type_gearbox=crosstab_function(data_type_car_gearbox,"brand")

In [121]:
data_brand_type_gearbox

brand  Automatica  Manual  All
0       Abarth           3       6    9
1         Alfa           1       3    4
2         Audi          21      23   44
3          BMW          34      23   57
4    Chevrolet           0       1    1
5      Citroen           3      39   42
6        Cupra           1       1    2
7           DS           0       1    1
8        Dacia           0      14   14
9         Fiat           1      21   22
10        Ford           6      54   60
11       Honda           0       6    6
12     Hyundai           1      22   23
13      Jaguar           6       0    6
14        Jeep           5       7   12
15         Kia           7      32   39
16        Land           8       3   11
17       Lexus           3       0    3
18       Mazda           5      13   18
19    Mercedes          23      13   36
20        Mini           8      12   20
21  Mitsubishi           2       7    9
22      Nissan           7      30   37
23        Opel           1      30   31
24     Peugeot          10      38   48
25     Porsche           4       0    4
26     Renault           7      32   39
27        SEAT          12      72   84
28       Skoda           0       2    2
29       Smart           2       1    3
30   Ssangyong           0       2    2
31      Subaru           2       0    2
32      Suzuki           1       3    4
33      Toyota          25      17   42
34  Volkswagen          23      28   51
35       Volvo           1       3    4
36         All         233     559  792

In [122]:
import plotly.express as px

In [123]:
fig = px.bar(data_brand_type_gearbox[0:36], x='brand', y=['Automatica',"Manual"])
fig.show()

The brand have most manual gearbox is `SEAT`
<br> `Chevrolet` and `DS` just have one 1 manual gearbox
<br> `Porsche` just have gearbox automatic

### Statistical describe

In [124]:
data_car["Price_Rp"].describe()

count    7.920000e+02
mean     2.580820e+08
std      1.128460e+08
min      8.062343e+07
25%      1.834627e+08
50%      2.421934e+08
75%      2.987429e+08
max      1.334245e+09
Name: Price_Rp, dtype: float64

In [125]:
data_car.describe()

price (eur)         year  mileage (kms)      Price_Rp
count    792.000000   792.000000     792.000000  7.920000e+02
mean   15973.383838  2016.404040   83228.833333  2.580820e+08
std     6984.338370     2.452251   46202.118589  1.128460e+08
min     4990.000000  2004.000000    1258.000000  8.062343e+07
25%    11355.000000  2015.000000   47145.500000  1.834627e+08
50%    14990.000000  2017.000000   80065.500000  2.421934e+08
75%    18490.000000  2018.000000  116321.750000  2.987429e+08
max    82580.000000  2021.000000  220000.000000  1.334245e+09

In [126]:
data_car.describe(include='object')

brand model                     engine    fuel gearbox    location
count    792   792                        792     792     792         792
unique    36   206                        694       5       2          10
top     SEAT  Leon  1.0 Ecoboost 92kW ST-Line  Diésel  Manual  Sabadell 1
freq      84    32                          8     377     559         123

Brand most in this data is `SEAT` have 84 frequency  <br>
gearbox most in this data is `Manual` have 559 frequency  <br>


# Clustering

In [127]:
data_car["location"].unique()	

array(['Granollers', 'Viladecans', 'Sabadell 1', 'Gavá', 'Mataró',
       'Sabadell 2', 'Esplugas', 'Hospitalet', 'Badalona', 'Sabadell 3'],
      dtype=object)

In [128]:
data_car["year"].unique()	

array([2016, 2014, 2011, 2018, 2017, 2020, 2015, 2019, 2012, 2021, 2013,
       2010, 2007, 2008, 2009, 2004])

In [129]:
data_car_for_clustering=data_car.drop(columns=(["year","price (eur)"]))

## one hot encoding

In [130]:
data_car_for_clustering.head(3)

brand    model                 engine  mileage (kms)      fuel  \
0     SEAT    Ibiza  SC 1.2 TSI 90cv Style          67000  Gasolina   
1  Hyundai      i30   1.6 CRDi 110cv Tecno         104868    Diésel   
2      BMW  Serie 5           530d Touring         137566    Diésel   

      gearbox    location   Price_Rp  
0      Manual  Granollers  145251430  
1      Manual  Viladecans  161408430  
2  Automatica  Viladecans  217957930

In [131]:
data_car_for_clustering=pd.get_dummies(data_car_for_clustering, columns=["brand","model","engine","fuel","gearbox","location"])

In [132]:
data_car_for_clustering

mileage (kms)   Price_Rp  brand_Abarth  brand_Alfa  brand_Audi  \
0            67000  145251430             0           0           0   
1           104868  161408430             0           0           0   
2           137566  217957930             0           0           0   
3            44495  403763430             0           0           0   
4            69800  169002220             0           0           0   
..             ...        ...           ...         ...         ...   
787          84040  226036430             0           0           0   
788          65872  403763430             0           0           0   
789          28830  290664430             1           0           0   
790         162895  242193430             0           0           0   
791           9800  232499230             0           0           0   

     brand_BMW  brand_Chevrolet  brand_Citroen  brand_Cupra  brand_DS  ...  \
0            0                0              0            0         0  ...   
1            0                0              0            0         0  ...   
2            1                0              0            0         0  ...   
3            0                0              0            0         0  ...   
4            0                0              0            0         0  ...   
..         ...              ...            ...          ...       ...  ...   
787          0                0              0            0         0  ...   
788          0                0              0            0         0  ...   
789          0                0              0            0         0  ...   
790          0                0              0            0         0  ...   
791          0                0              0            0         0  ...   

     location_Badalona  location_Esplugas  location_Gavá  location_Granollers  \
0                    0                  0              0                    1   
1                    0                  0              0                    0   
2                    0                  0              0                    0   
3                    0                  0              0                    0   
4                    0                  0              0                    0   
..                 ...                ...            ...                  ...   
787                  0                  0              1                    0   
788                  0                  0              0                    0   
789                  0                  0              0                    0   
790                  0                  0              0                    0   
791                  1                  0              0                    0   

     location_Hospitalet  location_Mataró  location_Sabadell 1  \
0                      0                0                    0   
1                      0                0                    0   
2                      0                0                    0   
3                      0                0                    0   
4                      0                0                    1   
..                   ...              ...                  ...   
787                    0                0                    0   
788                    0                0                    0   
789                    0                1                    0   
790                    0                1                    0   
791                    0                0                    0   

     location_Sabadell 2  location_Sabadell 3  location_Viladecans  
0                      0                    0                    0  
1                      0                    0                    1  
2                      0                    0                    1  
3                      0                    0                    1  
4                      0                    0                    0  
..                   ...           

In [140]:
colums=[]
for v in data_car_for_clustering.columns:
  colums.append(v)

In [133]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

In [134]:
scaler = StandardScaler()

In [135]:
scaler.fit(data_car_for_clustering)

StandardScaler()

In [136]:
df_scaled = scaler.transform(data_car_for_clustering)

In [141]:
df_scaled=pd.DataFrame(df_scaled, columns=[colums])

In [142]:
df_scaled

mileage (kms)  Price_Rp brand_Abarth brand_Alfa brand_Audi brand_BMW  \
0       -0.351479 -1.000495    -0.107211  -0.071247  -0.242536 -0.278480   
1        0.468655 -0.857227    -0.107211  -0.071247  -0.242536 -0.278480   
2        1.176818 -0.355789    -0.107211  -0.071247  -0.242536  3.590924   
3       -0.838886  1.291792    -0.107211  -0.071247  -0.242536 -0.278480   
4       -0.290838 -0.789891    -0.107211  -0.071247  -0.242536 -0.278480   
..            ...       ...          ...        ...        ...       ...   
787      0.017568 -0.284155    -0.107211  -0.071247  -0.242536 -0.278480   
788     -0.375909  1.291792    -0.107211  -0.071247  -0.242536 -0.278480   
789     -1.178154  0.288916     9.327379  -0.071247  -0.242536 -0.278480   
790      1.725386 -0.140887    -0.107211  -0.071247  -0.242536 -0.278480   
791     -1.590300 -0.226848    -0.107211  -0.071247  -0.242536 -0.278480   

    brand_Chevrolet brand_Citroen brand_Cupra  brand_DS  ...  \
0         -0.035556     -0.236643   -0.050315 -0.035556  ...   
1         -0.035556     -0.236643   -0.050315 -0.035556  ...   
2         -0.035556     -0.236643   -0.050315 -0.035556  ...   
3         -0.035556     -0.236643   -0.050315 -0.035556  ...   
4         -0.035556     -0.236643   -0.050315 -0.035556  ...   
..              ...           ...         ...       ...  ...   
787       -0.035556     -0.236643   -0.050315 -0.035556  ...   
788       -0.035556     -0.236643   -0.050315 -0.035556  ...   
789       -0.035556     -0.236643   -0.050315 -0.035556  ...   
790       -0.035556     -0.236643   -0.050315 -0.035556  ...   
791       -0.035556     -0.236643   -0.050315 -0.035556  ...   

    location_Badalona location_Esplugas location_Gavá location_Granollers  \
0           -0.330520         -0.313806     -0.301511            3.186680   
1           -0.330520         -0.313806     -0.301511           -0.313806   
2           -0.330520         -0.313806     -0.301511           -0.313806   
3           -0.330520         -0.313806     -0.301511           -0.313806   
4           -0.330520         -0.313806     -0.301511           -0.313806   
..                ...               ...           ...                 ...   
787         -0.330520         -0.313806      3.316625           -0.313806   
788         -0.330520         -0.313806     -0.301511           -0.313806   
789         -0.330520         -0.313806     -0.301511           -0.313806   
790         -0.330520         -0.313806     -0.301511           -0.313806   
791          3.025532         -0.313806     -0.301511           -0.313806   

    location_Hospitalet location_Mataró location_Sabadell 1  \
0              -0.33052       -0.321036           -0.428785   
1              -0.33052       -0.321036           -0.428785   
2              -0.33052       -0.321036           -0.428785   
3              -0.33052       -0.321036           -0.428785   
4              -0.33052       -0.321036            2.332172   
..                  ...             ...                 ...   
787            -0.33052       -0.321036           -0.428785   
788            -0.33052       -0.321036           -0.428785   
789            -0.33052        3.114916           -0.428785   
790            -0.33052        3.114916           -0.428785   
791            -0.33052       -0.321036           -0.428785   

    location_Sabadell 2 location_Sabadell 3 location_Viladecans  
0             -0.293972            -0.33052           -0.358057  
1             -0.293972            -0.33052            2.792848  
2             -0.293972            -0.33052            2.792848  
3             -0.293972            -0.33052            2.792848  
4             -0.293972            -0.33052           -0.358057  
..                  ...                 ...                 ...  
787           -0.293972            -0.33052           -0.358057  
788           -0.293972            -0.33052            2.792848  
789           -0.293972            -0.33052   